py/server/deephaven/column.py

```diff
@@ -32,47 +32,105 @@ def __repr__(self):
+                )
+            self.name = j_column_definition.getName()
+            self.j_column_definition = j_column_definition
+        else:
+            if not name or not data_type:
+                raise DHError("Must specify both name and data_type")
+            if hasattr(data_type.j_type, "jclass"):
+                j_data_type = data_type.j_type.jclass
+            else:
+                j_data_type = data_type.qst_type.clazz()
+            j_component_type = (
+                component_type.qst_type.clazz() if component_type else None
+            )
+            j_column_type = column_type.value
+            self.name = name
+            self.j_column_definition = _JColumnDefinition.fromGenericType(
+                name, j_data_type, j_component_type, j_column_type
+            )
 
     @property
-    def j_column_header(self):
-        return _JColumnHeader.of(self.name, self.data_type.qst_type)
+    def j_object(self) -> jpy.JType:
+        return self.j_column_definition
 
-    @property
-    def j_column_definition(self):
-        if hasattr(self.data_type.j_type, 'jclass'):
-            j_data_type = self.data_type.j_type.jclass
-        else:
-            j_data_type = self.data_type.qst_type.clazz()
-        j_component_type = self.component_type.qst_type.clazz() if self.component_type else None
-        j_column_type = self.column_type.value
-        return _JColumnDefinition.fromGenericType(self.name, j_data_type, j_component_type, j_column_type)
+    @cached_property
+    def data_type(self) -> DType:
+        return from_jtype(self.j_column_definition.getDataType())
 
+    @cached_property
+    def component_type(self) -> DType:
+        return from_jtype(self.j_column_definition.getComponentType())
 
-@dataclass
-class InputColumn(Column):
-    """ An InputColumn represents a user defined column with some input data. """
-    input_data: Any = field(default=None)
+    @cached_property
+    def column_type(self) -> ColumnType:
+        return ColumnType(self.j_column_definition.getColumnType())
 
-    def __post_init__(self):
+    @property
+    def j_column_header(self):
+        return _JColumnHeader.of(self.name, self.data_type.qst_type)
+
+    def _to_j_column(self, input_data: Any = None):
+        if input_data is None:
+            return _JColumn.empty(self.j_column_header)
+        if self.data_type.is_primitive:
+            return _JColumn.ofUnsafe(
+                self.name,
+                dtypes.array(
+                    self.data_type,
+                    input_data,
+                    remap=dtypes.null_remap(self.data_type),
+                ),
+            )
+        return _JColumn.of(
+            self.j_column_header,
+            dtypes.array(self.data_type, input_data),
+        )
+
+
+class InputColumn(JObjectWrapper):
+    """ An InputColumn represents a user defined column with some input data."""
```

Comment:
1. I don't know if `InputColumn` needs to be a wrapper. 
2. We could change the relationship between `InputColumn` and 'Column` from subclassing to containment, and still expose the attributes of `Column` for `InputColumn`, and thus maintain backwards compatibility
3. I would move the `_to_j_column` from `Column` to `InputColumn`
4. `_to_j_column` could use a return type hint of `jpy.JType`

Reply:
Done